### ISSUE TITLE
How to Do Code Reviews Like a Human (Part Two)

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#7390]
How to Do Code Reviews Like a Human (Part Two)<br>
<br>
https://ift.tt/2yLZSOY<br>
<br>
Michael Lynch<br>
<br>
<div>
<div>
<div>
<div>
<p>This is the second half of my article about how to communicate well and avoid pitfalls in code reviews. Here, I focus on techniques to bring your code review to a successful close while avoiding ugly conflict.</p>
<p>I laid the groundwork in <a href="https://mtlynch.io/human-code-reviews-1/">Part One</a>, so I recommend starting there. If you&rsquo;re impatient, here&rsquo;s the short version: a good code reviewer not only finds bugs but provides conscientious feedback to help their teammates improve.</p>
<h2>My worst code review <a href="https://mtlynch.io/human-code-reviews-2/#my-worst-code-review">&#128279;&#65038;</a>
</h2>
<p>The worst code review of my life was for a former teammate I&rsquo;ll call Mallory. She started at the company several years before I joined but had only recently transferred to my team.</p>
<h3>The review <a href="https://mtlynch.io/human-code-reviews-2/#the-review">&#128279;&#65038;</a>
</h3>
<p>When Mallory sent me her first changelist for review, the code was a bit rough. She had never written Python before, and she was building on top of a clunky, legacy system that I maintained.</p>
<p>I dutifully recorded all of the issues I spotted, 59 in total. According to the review literature I&rsquo;d read, I had done a great job. I found SO many mistakes. Therefore, I must be a good reviewer.</p>
<p>A few days later, Mallory sent me the updated changelist and her responses to my notes. She had fixed the simple issues: typos, variable renames, etc. But she refused to address the higher-level problems, such as the fact that her code had undefined behavior for malformed input or that one of her functions nested control-flow structures six layers deep. Instead, she explained dismissively that these issues were not worth the engineering time to fix.</p>
<p>Angry and frustrated, I sent a new round of notes. My tone was professional but meandering into the realm of passive-aggressive. &ldquo;Can you explain <em>why</em> we want undefined behavior for malformed input?&rdquo; As you might guess, Mallory&rsquo;s replies became even more obstinate.</p>
<h3>A bitter cycle <a href="https://mtlynch.io/human-code-reviews-2/#a-bitter-cycle">&#128279;&#65038;</a>
</h3>
<p>It was Tuesday, a week later. Mallory and I were still going back and forth on the same review. I had sent her my latest notes the evening before. I purposely withheld them until she left for the day because I didn&rsquo;t want to be in the same room when she read them.</p>
<p>All morning, I felt a sinking weight in the pit of my stomach as I dreaded the next round of review. I came back from lunch to see that Mallory was away from her desk but had sent me new changes. I guess she didn&rsquo;t want to be around to see me read her replies either.</p>
<p>My heart began pounding in my chest as I grew more infuriated by each of her responses. I immediately started hammering my keyboard with rebuttals, pointing out that she had neither made my suggested changes nor offered justification for me to approve.</p>
<p>We repeated this routine every day for three weeks. The code barely changed.</p>
<h3>Intervention <a href="https://mtlynch.io/human-code-reviews-2/#intervention">&#128279;&#65038;</a>
</h3>
<p>Our most senior teammate, Bob, thankfully broke this cycle. He returned from a long vacation, alarmed to find us bitterly flinging code review notes back and forth. He immediately recognized the situation for what it was: a stalemate. He requested to take over the review, and we both agreed.</p>
<p>Bob began his review by asking Mallory to create new changelists, splitting off two small libraries that we had never really fought about, each about 30-50 lines. Once Mallory did that, Bob instantly approved them.</p>
<p>Then, Bob came back to the main changelist, which was trimmed down to about 200 lines of code. He made a few small suggestions, which Mallory addressed. Then, he approved the changelist.</p>
<p>Bob&rsquo;s entire review was done in two days.</p>
<h3>Communication matters <a href="https://mtlynch.io/human-code-reviews-2/#communication-matters">&#128279;&#65038;</a>
</h3>
<p>You may have deduced that this conflict wasn&rsquo;t really about the code. It had legitimate issues, but they were clearly solvable by teammates who could communicate effectively.</p>
<p>It was an unpleasant experience, but one I&rsquo;m glad for in retrospect. It caused me to reevaluate my approach to reviews and identify areas for improvement.</p>
<p>Below, I share techniques that will reduce your risk of a similarly undesirable outcome. I&rsquo;ll return to Mallory later and explain why my original approach was backward and why Bob&rsquo;s was quietly brilliant.</p>
<h2>Techniques <a href="https://mtlynch.io/human-code-reviews-2/#techniques">&#128279;&#65038;</a>
</h2>
<ol>
<li><a href="https://mtlynch.io/human-code-reviews-2/#aim-to-bring-the-code-up-a-letter-grade-or-two">Aim to bring the code up a letter grade or two</a></li>
<li><a href="https://mtlynch.io/human-code-reviews-2/#limit-feedback-on-repeated-patterns">Limit feedback on repeated patterns</a></li>
<li><a href="https://mtlynch.io/human-code-reviews-2/#respect-the-scope-of-the-review">Respect the scope of the review</a></li>
<li><a href="https://mtlynch.io/human-code-reviews-2/#look-for-opportunities-to-split-up-large-reviews">Look for opportunities to split up large reviews</a></li>
<li><a href="https://mtlynch.io/human-code-reviews-2/#offer-sincere-praise">Offer sincere praise</a></li>
<li><a href="https://mtlynch.io/human-code-reviews-2/#grant-approval-when-remaining-fixes-are-trivial">Grant approval when remaining fixes are trivial</a></li>
<li><a href="https://mtlynch.io/human-code-reviews-2/#handle-stalemates-proactively">Handle stalemates proactively</a></li>
</ol>
<h3>Aim to bring the code up a letter grade or two <a href="https://mtlynch.io/human-code-reviews-2/#aim-to-bring-the-code-up-a-letter-grade-or-two">&#128279;&#65038;</a>
</h3>
<p>While your teammate might, in <em>theory</em>, want to explore every opportunity to improve their code, their patience is finite. They&rsquo;ll quickly grow frustrated if you withhold approval round after round because you keep thinking of new and brilliant ways for them to polish their changelist.</p>
<p>I privately think of the code in terms of letter grades, from A to F. When I receive a changelist that starts at a D, I try to help the author bring it to a C or a B-. Not perfect, but good enough.</p>
<p>It&rsquo;s possible, in theory, to bring a D up to an A+, but it will probably take upwards of eight rounds of review. By the end, the author will hate you and never want to send you code again.</p>
<p>You might be thinking, &ldquo;If I accept C-grade code, won&rsquo;t I end up with a C-grade codebase?&rdquo; Fortunately, no. I find that when I help a teammate go from a D to a C, the next changelist they send me will <em>start</em> at a C. Within a few months, they&rsquo;re sending me reviews that begin as Bs, which become As by the end of the review.</p>
<p>An F is reserved for code that is either functionally incorrect or so convoluted that you don&rsquo;t have confidence in its correctness. The only reason you should withhold approval is if the code remains at an F after a few rounds of review. See the section on <a href="https://mtlynch.io/human-code-reviews-2/#handle-stalemates-proactively">stalemates</a>, below.</p>
<h3>Limit feedback on repeated patterns <a href="https://mtlynch.io/human-code-reviews-2/#limit-feedback-on-repeated-patterns">&#128279;&#65038;</a>
</h3>
<p>When you notice that several of the author&rsquo;s mistakes fit the same pattern, don&rsquo;t flag every single instance. You don&rsquo;t want to spend your time writing the same note 25 times, and the author certainly doesn&rsquo;t want to read 25 duplicate notes.</p>
<p>It&rsquo;s fine to call out two or three separate instances of a pattern. For anything more than that, just ask the author to fix the pattern rather than each particular occurrence.</p>
<h3>Respect the scope of the review <a href="https://mtlynch.io/human-code-reviews-2/#respect-the-scope-of-the-review">&#128279;&#65038;</a>
</h3>
<p>There&rsquo;s an anti-pattern I see frequently where the reviewer identifies something <em>near</em> code in the changelist and asks the author to fix it. Once the author complies, the reviewer usually realizes that the code is better but inconsistent, so it needs a few more minor changes. And then a few more. And on and on until a narrowly-scoped changelist has expanded to include lots of unrelated churn.</p>
<blockquote>
<p>If a hungry little mouse shows up on your doorstep, you might want to give him a cookie. And if you give him a cookie, he&rsquo;ll ask for a glass of milk. He&rsquo;ll want to look in a mirror to make sure he doesn&rsquo;t have a milk mustache, and then he&rsquo;ll ask for a pair of scissors to give himself a trim&hellip;</p>
<p>-Laura Joffe Numeroff, <a href="http://amzn.to/2iFNk4u"><em>If You Give a Mouse a Cookie</em></a></p>
</blockquote>
<p>The rule of thumb is: if the changelist doesn&rsquo;t touch the line, it&rsquo;s out of scope.</p>
<p>Here&rsquo;s an example:</p>
<p>Even if you&rsquo;ll be kept awake all night, haunted by the <a href="https://en.wikipedia.org/wiki/Magic_number_(programming)">magic number</a> and ridiculous variable name in your codebase, it&rsquo;s out of scope. Even if the author is the same person who wrote the nearby lines, it&rsquo;s still out of scope. If it&rsquo;s egregiously bad, file a bug or submit your own fix, but don&rsquo;t force it onto the author&rsquo;s plate in this review.</p>
<p>The exception is when the changelist affects the surrounding code without actually touching it, for example:</p>
<p>In this case, point out that the author needs to rename the function from <code>ValidateAndSerialize</code> to just <code>Serialize</code>. They haven&rsquo;t touched the line containing the function signature, but they still caused it to become incorrect.</p>
<p>I softly break this rule if I don&rsquo;t have many notes but notice an easy fix just out of scope. In these cases, I make it clear that the author can ignore the note if they please.</p>
<h3>Look for opportunities to split up large reviews <a href="https://mtlynch.io/human-code-reviews-2/#look-for-opportunities-to-split-up-large-reviews">&#128279;&#65038;</a>
</h3>
<p>If you receive a changelist that&rsquo;s more than ~400 lines of code, encourage the author to split it into smaller pieces. Push back proportionally harder the more they go over this limit. I personally refuse to review any changelists that exceed 1,000 lines.</p>
<p>The author may gripe about splitting the changelist because it&rsquo;s a tedious task. Ease their burden by identifying logical boundaries for the split. The easiest case is when the changelist touches multiple files independently. In that case, they can just split the changelist into smaller sets of files. In harder cases, find the functions or classes at the lowest layer of abstraction. Ask the author to move these to a separate changelist, then circle back to the rest of the code after the first changelist is merged in.</p>
<p>When the code quality is low, <em>emphatically</em> request a split. The difficulty of reviewing bad code grows exponentially with size. You&rsquo;re much better off auditing a couple of sloppy 300-line changelists than a single 600-line abomination.</p>
<h3>Offer sincere praise <a href="https://mtlynch.io/human-code-reviews-2/#offer-sincere-praise">&#128279;&#65038;</a>
</h3>
<p>Most reviewers focus only on what&rsquo;s <em>wrong</em> with the code, but reviews are a valuable opportunity to reinforce positive behaviors.</p>
<p>For example, imagine you&rsquo;re reviewing for an author who struggles to write documentation, and you come across a clear, concise function comment. Let them know they nailed it. They&rsquo;ll improve faster if you tell them when they got it right instead of just waiting to ding them when they screw up.</p>
<p>You don&rsquo;t need to have a specific goal in mind to offer praise. Any time I see something in the changelist that delights me, I tell the author about it:</p>
<ul>
<li>&ldquo;I wasn&rsquo;t aware of this API. That&rsquo;s really useful!&rdquo;</li>
<li>&ldquo;This is an elegant solution. I never would have thought of that.&rdquo;</li>
<li>&ldquo;Breaking up this function was a great idea. It&rsquo;s so much simpler now.&rdquo;</li>
</ul>
<p>If the author is a junior developer or joined the team recently, they&rsquo;re likely to feel nervous or defensive during a review. Sincere compliments ease this tension by demonstrating that you are their supportive teammate and not the cruel gatekeeper.</p>
<h3>Grant approval when remaining fixes are trivial <a href="https://mtlynch.io/human-code-reviews-2/#grant-approval-when-remaining-fixes-are-trivial">&#128279;&#65038;</a>
</h3>
<p>Some reviewers have the misconception that they should withhold approval until they witness fixes for every last note. This adds needless code review rounds, wasting time for both author and reviewer.</p>
<p>Grant approval when any of the following are true:</p>
<ul>
<li>You have no more notes.</li>
<li>Your remaining notes are for trivial issues.
<ul>
<li>E.g., renaming a variable, fixing a typo</li>
</ul>
</li>
<li>Your remaining notes are optional suggestions.
<ul>
<li>Explicitly mark these as optional so that your teammate doesn&rsquo;t assume your approval is contingent on them.</li>
</ul>
</li>
</ul>
<p>I&rsquo;ve seen reviewers withhold approval because the author missed a period at the end of a code comment. Please don&rsquo;t do this. It signals to the author that you think they&rsquo;re incapable of adding simple punctuation unless supervised.</p>
<p>There is some danger in granting approval when there are still outstanding notes. I estimate that ~5% of the time, the author either misinterprets a final round note or misses it completely. To mitigate this, I simply check the author&rsquo;s post-approval changes. In the rare case of miscommunication, I either follow up with the author or create my own changelist with a fix. Adding a small amount of work to the 5% case is better than adding unnecessary effort and delay to other 95%.</p>
<h3>Handle stalemates proactively <a href="https://mtlynch.io/human-code-reviews-2/#handle-stalemates-proactively">&#128279;&#65038;</a>
</h3>
<p>The worst possible outcome of a code review is a stalemate: you refuse to sign off on the changelist without further changes, but the author refuses to make them.</p>
<p>Here are some indicators that you&rsquo;re headed for a stalemate:</p>
<ul>
<li>The tone of the discussion is growing tense or hostile.</li>
<li>Your notes per review round are not trending downward.</li>
<li>You&rsquo;re getting pushback on an unusually high number of your notes.</li>
</ul>
<h3>Talk it out <a href="https://mtlynch.io/human-code-reviews-2/#talk-it-out">&#128279;&#65038;</a>
</h3>
<p>Meet in person or over video chat. Text communication has a way of causing you to forget there&rsquo;s a real human at the other end of the conversation. It becomes too easy to imagine your teammate is coming from a place of stubbornness or incompetence. A meeting will break that spell for both you and the author.</p>
<h3>Consider a design review <a href="https://mtlynch.io/human-code-reviews-2/#consider-a-design-review">&#128279;&#65038;</a>
</h3>
<p>A contentious code review may indicate weaknesses earlier in the process. Are you arguing about things that should have been covered during the design review? <em>Was</em> there a design review?</p>
<p>If the root of the disagreement traces back to a high-level design choice, the broader team should weigh in rather than leave it in the hands of the two people who happen to be in the code review. Talk to the author about opening up the discussion to the rest of your team in the form of a design review.</p>
<h3>Concede or Escalate <a href="https://mtlynch.io/human-code-reviews-2/#concede-or-escalate">&#128279;&#65038;</a>
</h3>
<p>The longer you and your teammate stew in stalemate, the more damaging it is to your relationship. If alternatives haven&rsquo;t gotten you unstuck, your options are to either concede or escalate.</p>
<p>Weigh the cost of just approving the changes. You can&rsquo;t build quality software if you casually accept low-quality code, but you also can&rsquo;t achieve high quality when you and your teammate fight so bitterly that you can no longer work together. How bad would it <em>really</em> be if you approved the changelist? Is it code that could potentially destroy critical data? Or is it a background process where, at worst, the job will fail and require a developer to debug it? If it&rsquo;s closer to the latter, consider simply conceding so that you can continue working with your teammate on good terms.</p>
<p>If concession is not an option, talk to the author about escalating the discussion to your team&rsquo;s manager or tech lead. Offer to reassign to a different reviewer. If the escalation goes against you, accept the decision and move on. Continuing to fight it will drag out a bad situation and make you look unprofessional.</p>
<h3>Recovering from a stalemate <a href="https://mtlynch.io/human-code-reviews-2/#recovering-from-a-stalemate">&#128279;&#65038;</a>
</h3>
<p>Messy review arguments tend to be less about the code and more about the relationship between the people involved. If you reached stalemate or near-stalemate, this pattern will repeat if you don&rsquo;t address the underlying conflict.</p>
<ul>
<li>Discuss the situation with your manager.
<ul>
<li>If there&rsquo;s conflict on the team, your manager should know about it. Maybe the author is just difficult to work with. Perhaps you&rsquo;re contributing to the situation in ways you don&rsquo;t recognize. A good manager will help both of you address these issues.</li>
</ul>
</li>
<li>Take a break from each other.
<ul>
<li>If possible, avoid sending each other code reviews for a few weeks until things cool down.</li>
</ul>
</li>
<li>Study conflict resolution.
<ul>
<li>I found the book <a href="http://amzn.to/2hvUbsP"><em>Crucial Conversations</em></a> to be helpful. Its advice may sound common-sense, but there&rsquo;s tremendous value in analyzing your approach to conflict while you&rsquo;re not in the heat of an argument.</li>
</ul>
</li>
</ul>
<h2>My worst code review: revisited <a href="https://mtlynch.io/human-code-reviews-2/#my-worst-code-review-revisited">&#128279;&#65038;</a>
</h2>
<p>Remember the code review with Mallory? Why did mine turn into a three-week slog through passive-aggressive muck while Bob&rsquo;s was a two-day breeze?</p>
<h3>What I did wrong <a href="https://mtlynch.io/human-code-reviews-2/#what-i-did-wrong">&#128279;&#65038;</a>
</h3>
<p>This was Mallory&rsquo;s first review on the team. I failed to consider that she might feel judged or defensive. I should have <a href="https://mtlynch.io/human-code-reviews-1/#start-high-level-and-work-your-way-down">started out with only high-level comments</a> so that she didn&rsquo;t feel ambushed by the large volume of notes.</p>
<p>I should have done more to demonstrate that my job wasn&rsquo;t to obstruct her work, but rather help it move forward. I could have provided <a href="https://mtlynch.io/human-code-reviews-1/#be-generous-with-code-examples">code examples</a> or <a href="https://mtlynch.io/human-code-reviews-2/#offer-sincere-praise">called out the positives</a> in her changelist.</p>
<p>I allowed <a href="https://mtlynch.io/human-code-reviews-1/#never-say-you">my ego</a> to affect the review. I had spent the past year nursing this old system back to health. Suddenly, there was a new person futzing with it, but she couldn&rsquo;t be bothered to take my concerns seriously? I took it as an affront, but that attitude was counterproductive. I should have maintained the objective mindset I try to bring to all of my reviews.</p>
<p>Finally, I allowed the stalemate to drag on too long. After a few rounds, it should have been clear to me that we weren&rsquo;t making meaningful progress. I should have <a href="https://mtlynch.io/human-code-reviews-2/#handle-stalemates-proactively">made a drastic change</a>, such as meeting in person to address the deeper conflict or escalating to our manager.</p>
<h3>What Bob did right <a href="https://mtlynch.io/human-code-reviews-2/#what-bob-did-right">&#128279;&#65038;</a>
</h3>
<p>Bob&rsquo;s first move of <a href="https://mtlynch.io/human-code-reviews-2/#look-for-opportunities-to-split-up-large-reviews">splitting up the review</a> was very effective. Recall that the review that had been stalled for three painful weeks. Suddenly, two pieces of code were merged in. This made both Mallory and Bob feel good because it established forward momentum. There were still issues with the remaining chunk, but it became a smaller, easier-to-manage changelist.</p>
<p>Bob <a href="https://mtlynch.io/human-code-reviews-2/#aim-to-bring-the-code-up-a-letter-grade-or-two">didn&rsquo;t try to strangle the review to perfection</a>. He likely recognized the same issues that I was screaming about but realized Mallory would be on the team awhile. His flexibility in the short-term positioned him to help Mallory improve quality in the long-term.</p>
<h2>Conclusion <a href="https://mtlynch.io/human-code-reviews-2/#conclusion">&#128279;&#65038;</a>
</h2>
<p>After I published the first half of this article, several readers took issue with the communication style I recommended. Some found it patronizing. Others worried that it was too indirect and risked miscommunication.</p>
<p>That feedback is reasonable and expected. One person may find a terse review comment to be brusque or rude. Another may judge the same comment as concise and efficient.</p>
<p>In reviewing code, you make many choices: what to focus on, how to frame feedback, when to approve. It&rsquo;s not important that you choose <em>my</em> options. Just recognize that there <em>are</em> options.</p>
<p>No one can hand you a recipe for a perfect review. The techniques that work best will depend on the code author&rsquo;s personality, your relationship with them, and your team&rsquo;s culture. Hone your approach by thinking critically about the outcomes of your code reviews. When you encounter tension, take a step back to evaluate why it happened. Pay attention to the quality of your reviews. If you feel unable to bring code up to your quality standards, think about what aspects of the review process are hindering you and how you can address them.</p>
<p>Good luck, and may your code reviews be human-like.</p>
<h2>Further Reading <a href="https://mtlynch.io/human-code-reviews-2/#further-reading">&#128279;&#65038;</a>
</h2>
<p>Dr. Karl Wiegers is the only author I found who gave due attention to the social factors of code reviews. He summarizes his views nicely in his article, <a href="http://www.processimpact.com/articles/humanizing_reviews.pdf">&ldquo;Humanizing Peer Reviews."</a> Written in 2002, its continued relevance demonstrates the long-term value of effective communication.</p>
<hr>
<p><em>This article was edited by <a href="https://www.samanthamasonfreelancer.com">Samantha Mason</a>. Illustrations by <a href="https://www.linkedin.com/in/lolo-ology/">Loraine Yow</a>. Thanks to <a href="https://twitter.com/global4g">@global4g</a> for providing valuable feedback on an early draft of this post.</em></p>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via mtlynch.io<br>
<br>
December 9, 2020 at 08:35PM
